### ISSUE TITLE
better autocomplete using tern.js

### ISSUE THREAD

**Issue by kevinb7** [Khan/live-editor#320]
I had started implementing better autocomplete in https://github.com/Khan/live-editor/pull/178, but I didn't get very far with it.  This morning I ran across a library called tern.js.

> Tern is a stand-alone, editor-independent JavaScript analyzer that can be used to improve the JavaScript integration of existing editors.

It supports the following things:

> - Autocompletion on variables and properties
> - Function argument hints
> - Querying the type of an expression
> - Finding the definition of something
> - Automatic refactoring

We probably don't care about automatic refactoring, but the others would be quite helpful.  Although we do provide some argument hints, these are only for "built-in" functions.  For more information see: http://ternjs.net/

**Comment by kevinb7** [comment]
I had a quick look at the source and under the hood it's using acorn to do the parsing.  I think we'd probably want to fork it and add an option to pass in the AST so we're not parsing more than we have to.  Since acorn and esprima produce essentially the same AST, I don't see this being an issue.

**Comment by Gigabyte-Giant** [comment]
Very neat!
Quick question, if implemented, would the autocomplete menu be shown on the `keyup` event (or similar), rather than having to do a keystroke?

**Comment by kevinb7** [comment]
I think we'd just integrate it with the existing autocomplete menu which appears to occur on ```keydown```.

**Comment by hzoo** [comment]
@kevinb7 You will probably need to do a transform like we're doing in `babel-eslint`: [`acorn-to-esprima.js`](https://github.com/babel/babel-eslint/blob/56815a4ecf4f306a7fc4810934adb06a2e3480c5/acorn-to-esprima.js#L5) if you are already using esprima. ( author of tern also made acorn)

**Comment by kevinb7** [comment]
@hzoo good to know.  I thought the ASTs were the same, is some of the metadata formatted differently?

**Comment by hzoo** [comment]
Yeah there's a few small differences - the token metadata is a bit different and some parts of the tree/metadata is different. We're goina try to make `acorn-to-esprima` its own module at some point.

Also there's actually an effort to standardize the ast called `estree` https://github.com/estree/estree (with people from esprima/acorn/etc) and a lot of the es6+ stuff has been decided jointly.

**Comment by kevinb7** [comment]
Looking forward to ```estree``` becoming standardized.

**Comment by hzoo** [comment]
Oh should of mentioned that ES6 is basically standardized already across the parsers with ESTree - Also https://github.com/babel/acorn-to-esprima is it's own module now
